src/embedded/sh/embedder.ts

```diff
@@ -0,0 +1,99 @@
+import type { Options } from "prettier";
+import { builders, utils } from "prettier/doc";
+import type { Embedder } from "../../types.js";
+import { preparePlaceholder, printTemplateExpressions } from "../utils.js";
+import { embeddedLanguage } from "./embedded-language.js";
+
+const { hardline, group, line, softline, indent } = builders;
+const { mapDoc } = utils;
+
+export const embedder: Embedder<Options> = async (
```

Comment:
For most of the plugins, they're basically the same. But there're some plugins that need custom tweaking. I'll see what I can do to avoid most of the redundant code.